src/hotspot/share/oops/objArrayKlass.inline.hpp

```diff
@@ -70,7 +70,8 @@ void ObjArrayKlass::oop_oop_iterate_elements_bounded(
 
 template <typename T, typename OopClosureType>
 void ObjArrayKlass::oop_oop_iterate(oop obj, OopClosureType* closure) {
-  assert (obj->is_array(), "obj must be array");
+  // In this assert, we cannot safely access the Klass* with compact headers.
+  assert (UseCompactObjectHeaders || obj->is_array(), "obj must be array");
```

Comment:
If we can't safely access the `Klass*` here, why is the call to `obj->klass()` below safe?

Reply:
Good question. This comment and assert can probably be removed (same for the similar comment/assert in TypeArrayKlass::oop_oop_iterate_impl(). Could be a left-over from a time when we had to deal with OM and/or stack-locks in the header.

Reply:
FWIW, I've been running tests with this assert restored (and the one in TypeArrayKlass) without hitting any problems.